File: BOJ/Class3/적록색약.py
import sys
# sys.stdin = open("BOJ/Class3/input.txt")
input = sys.stdin.readline

N = int(input())
grid = [["" for _ in range(N)] for _ in range(N)]

for i in range(N):
    data = input().split()
    for j in range(N):
        grid[i][j] = data[0][j]

def myprint(arr, n, m):
    for i in range(n):
        for j in range(m):
            print(arr[i][j], end=" ")
        print()

# myprint(grid, N, N)

#-- init()

dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]

def dfs(X, Y, visit):
    stack = []
    stack.append([X, Y])
    visit[Y][X] = True
    
    while stack:
        x, y = stack.pop()
        target = grid[y][x] #-- R
        #-- 상하좌우 탐색
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            #-- 범위 안에 있으면서 방문하지 않은 같은 색상에 노드
            if 0 <= nx < N and 0 <= ny < N and visit[ny][nx] == False and grid[ny][nx] == target:
                #-- 방문 체크 후 스택에 추가
                visit[ny][nx] = True
                stack.append([nx, ny])
        
        
    

#-- step1. 구역별로 그래프 탐색 시작

visit = [[False for _ in range(N)] for _ in range(N)]
notRedGreen = 0
RedGreen = 0
for i in range(N):
    for j in range(N):
        #-- 방문하지 않았던 곳을 방문하면서 체크
        if visit[i][j] == False:
            dfs(j, i, visit) #-- X, Y, 방문 배열
            notRedGreen +=1
print(notRedGreen, end= " ")

#-- step2. 그리드에 있는 적색을 녹색으로 모두 변경 후 다시 재 탐색
#-- visit 배열 재할당 + 적색을 녹색으로 

for i in range(N):
    for j in range(N):
        visit[i][j] = False
        if grid[i][j] == "R":
            grid[i][j] = "G"

for i in range(N):
    for j in range(N):
        if visit[i][j] == False:
            dfs(j, i, visit)
            RedGreen +=1
            
print(RedGreen)

File: BOJ/Gold/적록색약.java
import java.util.*;
import java.io.*;
public class 적록색약 {
    public static String grid[][];
    public static boolean visit[][];

    public static int dx[] = {0, 0, -1, 1};
    public static int dy[] = {-1, 1, 0, 0};

    public static class Point {
        int x;
        int y;
        public Point(int x, int y){
            this.x = x;
            this.y = y;
        }
    }
    public static void myprint(int n, BufferedWriter bw) throws Exception{
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                bw.write(grid[i][j] + " ");
            }
            bw.newLine();
        }
    }
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String input[] = br.readLine().split(" ");
        int n = Integer.parseInt(input[0]);

        grid = new String[n][n];
        visit = new boolean[n][n];
        for(int i=0; i<n; i++){
            input = br.readLine().split("");
            for(int j=0; j<n; j++){
                grid[i][j] = input[j];
            }
        }
        // myprint(n, bw);

        int normal = 0;
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(visit[i][j] == false){
                    normal++;
                    dfs(new Point(j, i), grid[i][j]);
                }
            }
        }

        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                visit[i][j] = false;
                if(grid[i][j].equals("R")){
                    grid[i][j] ="G";
                }
            }
                
        }

        int red = 0;
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(visit[i][j] == false){
                    red++;
                    dfs(new Point(j, i), grid[i][j]);
                }
            }
        }
        // myprint(n, bw);

        bw.write(String.valueOf(normal) + " ");
        bw.write(String.valueOf(red));
        bw.newLine();

        bw.flush();
        bw.close();


    }

    static void dfs(Point p, String target){
        visit[p.y][p.x] = true;
        Stack<Point>stack = new Stack<>();
        stack.push(p);

        while (!stack.isEmpty()) {
            Point point = stack.pop();

            for(int i=0; i<4; i++){
                int nx = point.x + dx[i];
                int ny = point.y + dy[i];

                if(isRange(nx, ny, grid.length) && visit[ny][nx] == false && grid[ny][nx].equals(target)){
                    visit[ny][nx] = true;
                    stack.push(new Point(nx, ny));
                }
            }
        }
    }

    static boolean isRange(int x, int y, int n){
        if(x < 0 || n <=x)
            return false;
        if(y < 0 || n <=y)
            return false;
        return true;
    }
}
